Use ViewComponents for Search Forms

fosterfarrell9 opened the pull request:
### Summary

This pull request introduces a new, centralized `SearchForm` ViewComponent system (see also #851) to replace the scattered and duplicated search form implementations across the application. Previously, search forms were implemented in various view folders with significant code duplication, making them difficult to maintain and extend. This PR partially resolves #838.

For more details, see this [README](https://github.com/MaMpf-HD/mampf/blob/refactor/search-views/app/frontend/_components/search_form/README.md) file.

### Possible Future Work

While this PR establishes the core `SearchForm` system, it also highlights opportunities for further architectural improvements across the application:

**1. Refactor Data-Fetching Logic:**
Several of the new field components are responsible for both querying data and formatting it for the view. This pattern of mixing data retrieval and presentation logic is not unique to these components and can be found in many other places in the app. The plan is to consistently apply a **Query Object / Presenter pattern** to address this, and the search filters will serve as the first step in this larger, application-wide initiative.

**2. Create a `SearchResults` Component:**
Currently, the presentation of search results is handled by various partials, leading to duplicated view logic. A logical next step is to create a `SearchResults` ViewComponent.

This component would follow a different design philosophy than the `SearchForm`. Instead of being a "strict" component that dictates content, it would be a "loose" layout component. It would handle the boilerplate "frame"—such as pagination, "no results" messages, and the collection loop—while yielding control back to the developer for rendering the unique content of each result row, likely via slots.

This approach would complement the `SearchForm` component, creating a fully standardized search experience from input to output, while respecting the unique presentation needs of different data models.

fosterfarrell9 commented on app/components/search_form/controls/radio_button.rb:
```diff
@@ -0,0 +1,69 @@
+      def initialize(value:, label:, checked: false, **options)
+        form_state = options.delete(:form_state)
+        name = options.delete(:name)
+
```
I agree. That's actually exactly what the code looked like before, but this resulted in a linter warning (too many parameters). But in this case, I think is justified to have these paramters and to have a clear API. Of course, one could further abstraction but iI just don't think it is worth it here.

fosterfarrell9 commented on app/components/search_form/services/filter_registry.rb:
```diff
@@ -0,0 +1,138 @@
+          klass.define_method("add_#{filter_name}_filter") do |**options|
+            filter = filter_class.constantize.new(**options)
+            with_field(filter)
+            # Return for testing
```
Fixed.

fosterfarrell9 commented on app/frontend/js/controllers/search_form_controller.js:
```diff
@@ -0,0 +1,137 @@
+  static targets = ["select", "allToggle", "radioToggle"];
+
+  connect() {
+    console.log("Search form controller connected");
```
Fixed.

Splines commented on app/components/search_form/README.md:
```diff
@@ -0,0 +1,261 @@
+# SearchForm Component System
+
+This document provides a guide to using and customizing the `SearchForm` component system.
```
The document does a great job explaining how things works at a high level.

I'm sure we can cut 30% of it by removing overly verbose language, sentences that don't provide any benefit for the understanding, and duplications (often disguised in something that is just rephrased). Here are just a few examples:

- No need to tell that something is "robust", I hope everything is robust here.
- Same for things like: "two-part system", when in the next two lines I see two bullet points.
- "Understanding X is essential for Y" (ok, but why not just show how it works, this is no prose but a technical document 😅).
- Also the "The DOM Traversal and Data Attribute Contract" is too detailed, e.g. how we traverse the DOM is not something for this high-level file I guess.
- Also a lot of duplications with regards to the scope in the customizing filter section. And I think we can cut the "Mechanism" sections completely.

Please also keep in mind that this file will age very quickly, especially all the filenames. So it might be a good idea to not include too many filenames here and instead just include class names etc. We can always use a quick file search to find the respective files.

fosterfarrell9 commented on .ruby-version:
That was an accidental commit that slipped in when I commited a larger batch of files. This file is created locally in my setting by rbenv. I would suggest that we add `.ruby-version` to the gitignore file in a different PR.

fosterfarrell9 commented on app/components/search_form/README.md:
I would probably still prefer `app/frontend/components/search_form` just to indicate that there are no actual views in this folder. I tried that but got stuck, Zeitwerk was no longer able to resolve the module structure inside the search form component. So I will leave that open, maybe you can try.

Splines replied:
fixed

fosterfarrell9 commented on spec/components/search_form/fields/utilities/group_wrapper_shared_spec.rb:
```diff
@@ -0,0 +1,135 @@
+      end
+
+      it "wraps the rendered items in a div if wrapper_class is provided" do
+        # rubocop:disable Rails/OutputSafety
+        wrapped_content = "<div class=\"wrapper\">#{joined_items}</div>".html_safe
+        # rubocop:enable Rails/OutputSafety
```
As before, it is a spec. Also, the `wrapped_content` variable is a string literal that represents the expected output of the `content_tag` method. We are using it to tell your mock what to return. What you suggest would be circular.

fosterfarrell9 commented on app/components/search_form/search_form.rb:
```diff
@@ -0,0 +1,109 @@
+    # @param context [String, nil] Context identifier for styling and behavior
+    # @param container_class [String] CSS classes for the main container div
+    # rubocop:disable Metrics/ParameterLists
+    def initialize(url:, scope: :search, method: :get, remote: true, context: nil,
+                   container_class: "row mb-3 p-2")
+      super()
+      @url = url
+      @scope = scope
+      @method = method
+      @remote = remote
+      @context = context || SecureRandom.hex(4)
+      @form_state = Services::FormState.new(context: @context)
+      @hidden_fields = {}
+      @container_class = container_class
```
See other comment.

fosterfarrell9 commented on app/components/search_form/fields/services/field_data.rb:
```diff
@@ -0,0 +1,182 @@
+        # This method intentionally accepts many parameters because this class acts as a
+        # Parameter Object. Each keyword argument represents a distinct, core configuration
+        # option for a field, and keeping them explicit improves API discoverability.
+        # rubocop:disable Metrics/ParameterLists
+        def initialize(name:, label:, form_state:, help_text: nil, options: {},
+                       multiple: nil, disabled: nil, required: nil, prompt: nil, selected: nil)
```
You suggest to use the parameter object pattern. Well, this *is* a parameter object:

```ruby
        # This method intentionally accepts many parameters because this class acts as a
        # Parameter Object. Each keyword argument represents a distinct, core configuration
        # option for a field, and keeping them explicit improves API discoverability.
```

This is a deliberate design decision.

fosterfarrell9 commented:
> The approach is really clever and the new view code looks gorgeous. But at the cost of a lot of complexity: by just going through some files, I quickly get a bit dizzy if I want to trace where data is passed to which other handler etc. A good indicator for whether it's a bit over-engineered or not is if you still remember how all files interact with each other after some weeks ;)

I agree that the first version was over-engineered. So I completely refactored. The new solution now uses composition over inheritance, there are no longer any complicated inheritance chains. The search form renders fields (placed in the `fields/`` subfolder), and each field inherits directly from `ViewComponent::base`. There are "primitive" fields in the `fields/primitives/` (text field, select field, ...) form which more complex fields can be composed from. Since composing a field comes with a certain amount of boilerplate code, I decided to use ruby mixins (in the `fields/mixins` folder) to keep thingy DRY. So if you want to write a new composite field, just include the composite field mixin, and life will be a little easier (or don't, and you'll have all the freedom you want). Also, in contrast to the previous solution, every single field now explicitly has its own separate template (explicit is better than implicit). I think this new solution is now a lot more developer friendly and maintainable.

Splines commented on app/views/search/_course_search_form.html.erb:
done in 950d59a716

Splines reviewed (changes requested):
- I moved the ViewComponents folder to `app/frontend/_components/` and fixed the Zeitwerk autloading for it (as usual this took me waaay too long haha, but at least now I have a better understanding of how this autloading works).
- I moved the search views to `app/frontend/search/`.
- I moved the JS Stimulus controller to `app/frontend/_components/search_form/`.

There are a few more threads open here that I left you ;) Could you also tackle the exception in the browser console that is likely related to this PR?

```sh
Uncaught (in promise) SyntaxError: Failed to execute 'querySelectorAll' on 'Document': 
'label[for="search_media_tag_operator_{:value=>"or"}"], label[for="search[tag_operator]"]' is not a valid selector.
```

The problem is likely the weird `_{:value=>"or"}` in the string.

---

Edit: Could you also update the PR description since I think it is outdated. For the detailed explanations, maybe just link to the Readme.

fosterfarrell9 commented on config/application.rb:
```diff
@@ -10,14 +10,15 @@ module Mampf
+    # General Zeitwerk Autoloading
+    backend_paths = Rails.root.glob("app/models/**/")
+    backend_paths -= Rails.root.glob("app/models/vignettes/**/")
+    frontend_paths = Rails.root.glob("app/frontend/**/")
```
Nice! Could you please add a comment that explains the logic behind this.

fosterfarrell9 commented on config/locales/de.yml:
```diff
@@ -2891,25 +2836,74 @@ de:
-        lassen. In eine eigene Auswahl kannst Du eine oder mehrere aufnehmen. Dabei wird der Operator
-        inklusives <i>ODER</i> verwendet.
-      lecture_scope:
+    fields:
```
That's good to know! Maybe we could at some point just write a little script that breaks our existing unwieldy locale files into smaller ones in subfolders? For this PR, though, I did it manually of course, and also discovered there were unused keys here which I removed then. Additionally, I changed the structure of the search locale files slightly (there was one duplicate layer related to fields) and replaced the references to them throughout the views. Also, I moved some keys that were previously sorted under `basics` but used only by the search forms to the search locale files. So, lots of fdile changes but not much that actually happened.

fosterfarrell9 commented:
> * I moved the ViewComponents folder to `app/frontend/_components/` and fixed the Zeitwerk autloading for it (as usual this took me waaay too long haha, but at least now I have a better understanding of how this autloading works).
> * I moved the search views to `app/frontend/search/`.
> * I moved the JS Stimulus controller to `app/frontend/_components/search_form/`.
> 
> There are a few more threads open here that I left you ;) Could you also tackle the exception in the browser console that is likely related to this PR?
> 
> ```shell
> Uncaught (in promise) SyntaxError: Failed to execute 'querySelectorAll' on 'Document': 
> 'label[for="search_media_tag_operator_{:value=>"or"}"], label[for="search[tag_operator]"]' is not a valid selector.
> ```
> 
> The problem is likely the weird `_{:value=>"or"}` in the string.
> 

I fixed that. Can you verify that the problem is gone now? In the process of resolving this bug I stumbled upon some additional minor things that needed fixing/refactoring, which I did.

> Edit: Could you also update the PR description? Since I think it is outdated. For the detailed explanations, maybe just link to the Readme.

I updated the PR description.

Splines commented on app/frontend/lectures/new/_new.html.erb:
```diff
@@ -61,10 +61,9 @@
         <%= f.select :sort,
-                     options_for_select(Lecture.sort_localized.invert.to_a,
-                                        'lecture'),
+                     options_for_select(Lecture.select_sorts, "lecture"),
```
Ah ok, thanks for clarifying, I think I'd prefer that the lectures controller just passes `sort: "lecture"` to the Lecture initializer.

Splines reviewed (dismissed):
I did not encounter the error anymore 👍 I played around with the search forms and everything seems to work nicely. The new structure with `.html.erb` files directly next to the respective `.rb` files is a lot better now. All in all, a real improvement over the code beforehand and the new view code looks gorgeous ;)

I have to admit that due to the size of this PR, I did not go through every line one by one as I usual do since this would take exponentially longer. But the unit tests seem to cover a lot and since the search works flawlessly (at least judging by my manual tests), I approve.
